### PR TITLE
Refs #32552 -- Added DiscoverRunner.log() to allow customization.

### PR DESCRIPTION
**Pull request by abbasidaniyal**
ticket-32552

Add a method `log(self, message, level)` to DiscoverRunner class which prints the message by comparing the level with self.verbosity. 
Replace the use of print with self.log

### CODE REVIEW

**Comment by abbasidaniyal** [comment]
The tests for DiscoverRunner
https://github.com/django/django/blob/37044817f9a57126d655f216019e8c8cca7c151b/tests/test_runner/test_discover_runner.py#L336-L337
and
https://github.com/django/django/blob/37044817f9a57126d655f216019e8c8cca7c151b/tests/test_runner/test_discover_runner.py#L252-L262

only check for verbosity >= 2. I believe we should add tests for verbosity < 2 as well. Thoughts?

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:584]: I had suggested on the ticket that this method accept a Python [logging level](https://docs.python.org/3/library/logging.html#logging-levels) like [`logging.log()`](https://docs.python.org/3/library/logging.html#logging.log), rather than an integer "verbosity." The logging level is the API that is most widely used and familiar to most people, and the goal of the ticket was to transition to using the logging module.

Looking at the [`--verbosity` docs](https://docs.djangoproject.com/en/3.1/ref/django-admin/#cmdoption-verbosity), I would say that a verbosity of 1 (the default) should show `INFO` and higher, while a verbosity of 2 or 3 (verbose or very verbose) should show `DEBUG`. The reason is that we want the ability to show some informational messages by default (like ticket #14137's shuffle seed). A verbosity of 0 should show no output. This logic can be reflected below.

**Review by cjerdonek — COMMENTED** [review]
Thanks. One comment..

**Comment by cjerdonek** [comment]
Also, the title of the PR should be something like "Added a DiscoverRunner.log() method" since you're not actually changing `DiscoverRunner` to use a logger here.

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:584]: Sure. Correct me if I'm wrong, 
We define constants like 
```
# Log Levels
NOTSET = 0
DEBUG = 10
INFO = 20
...
```

And in the __init__ method of DiscoverRunner, instead of setting self.verbosity = verbosity, we set self.level depending on the value of verbosity.
```
def __init__(self, pattern=None, top_level=None, verbosity=1,
                 interactive=True, failfast=False, keepdb=False,
                 reverse=False, debug_mode=False, debug_sql=False, parallel=0,
                 tags=None, exclude_tags=None, test_name_patterns=None,
                 pdb=False, buffer=False, enable_faulthandler=True,
                 timing=False, **kwargs):

        self.pattern = pattern
        self.top_level = top_level
        self.top_level = top_level
        if verbosity == 0:
            self.level = NOTSET
        elif verbosity == 1:
            self.level = INFO
        else:
            self.level = DEBUG
        self.interactive = interactive
```

Comment (cjerdonek) [reply django/test/runner.py:584]: > We define constants like

You don't need to define your own constants. You can use e.g. `logging.INFO`, etc. if you need to.

I would not modify `DiscoverRunner.__init__()`. Changes can be restricted to the new `DiscoverRunner.log()` using the current `self.verbosity` attribute.

**Comment by cjerdonek** [comment]
>  I believe we should add tests for verbosity < 2 as well. Thoughts?

I would add a single new test called `test_log()`. It can use `subTest()` to test the following matrix of 12 possibilities:  verbosity (0, 1, 2, 3) times level (DEBUG, INFO, WARN).

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:584]: I came across multiple instances where constants where manually defined. 
https://github.com/django/django/blob/41e6b2a3c5e723256506b9ff49437d52a1f3bf43/django/core/checks/messages.py#L2-L6
https://github.com/django/django/blob/41e6b2a3c5e723256506b9ff49437d52a1f3bf43/django/contrib/messages/constants.py#L1-L5
Just out of curiosity, any specific reasons for that? I personally think using logging.DEBUG etc is the better approach.

I modified `self.verbosity` because it is not used anywhere else. Seemed redundant to keep it as a class member.

Comment (cjerdonek) [reply django/test/runner.py:584]: > I modified self.verbosity because it is not used anywhere else. Seemed redundant to keep it as a class member.

I'm not sure what you mean. `self.verbosity` is used in a number of places in `DiscoverRunner`. `verbosity` is a documented argument, and subclassers (users of Django) can be using the attribute, too, so I think it should be kept.

Comment (lothemar) [reply django/test/runner.py:584]: +1, keep self.verbosity because third-party packages will assume ```self.verbosity``` is part of DiscoverRunner. This is a breaking change that's out of the scope of the PR.

**Comment by abbasidaniyal** [comment]
Just made few changes. Let me know if this seems better. Working on the tests!

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:39]: There's no need to define your own constants. You can just use `logging.INFO`, etc.

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:529]: You shouldn't modify `DiscoverRunner.__init__()`. All of the logic can go in `DiscoverRunner.log()`. In addition to not being necessary, a problem with setting `self.level` is that there will now no longer be one source of truth. For example, a caller that sets `DiscoverRunner.verbosity` after instantiating the runner will get incorrect results when calling `DiscoverRunner.log()`.

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:597]: The default level should be `logging.INFO` rather than `DEBUG`. However, I would implement this by making the default `None` (`level=None`) and interpreting `None` as `logging.INFO` in the body of the method.

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:643]: You won't need to pass `INFO` if the default is `INFO`.

**Review by cjerdonek — CHANGES_REQUESTED** [review]
> Just made few changes. Let me know if this seems better.

It looks like you didn't finish addressing my previous comments: https://github.com/django/django/pull/14150#discussion_r598203099

I've repeated them inline with some additional explanation and added a couple other comments.

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:39]: Oh sure. I mentioned that there are few other places where self-defined constants were used, so I followed that. Removing it!

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:529]: Sure. I have removed all the modifications in the __init__ and was planning on updating the defination of the `log` method as,
```
def log(self, msg, level = None):
        if level is None:
            level = logging.INFO
        
        if self.verbosity == 0:
            logging_level = logging.NOTSET
        elif self.verbosity == 1:
            logging_level = logging.INFO
        else:
            logging_level = logging.DEBUG

        if logging_level >= level:
            print(msg)
```

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:597]: Sure. Just out of curiosity, I wanted to ask, why would you implement it by defining `level = None` as the default value and the setting it to logging.INFO if level = None. Any specific reason or just a convention for better readability?

Comment (cjerdonek) [reply django/test/runner.py:597]: It simplifies getting the default behavior for callers because they can just pass `None`, instead of having to hard-code the default in a second location or invoke the function in a different way (without passing the argument).

**Review by carltongibson — CHANGES_REQUESTED** [review]
Hi @abbasidaniyal — Thanks for updating this. 

* Please look at the existing test failures. 
* Also see Chris' comment about adding a new `test_log()`, with the appropriate matrix of combinations. (See `subTest` Python docs and examples in the test suite for usage.) 

Once those are in place and addition to [the DiscoverRunner docs](https://docs.djangoproject.com/en/3.1/topics/testing/advanced#django.test.runner.DiscoverRunner) and a release note in `releases/4.0.txt` would be appropriate.

**Comment by abbasidaniyal** [comment]
> Hi @abbasidaniyal — Thanks for updating this.
> 
> * Please look at the existing test failures.
> * Also see Chris' comment about adding a new `test_log()`, with the appropriate matrix of combinations. (See `subTest` Python docs and examples in the test suite for usage.)
> 
> Once those are in place and addition to [the DiscoverRunner docs](https://docs.djangoproject.com/en/3.1/topics/testing/advanced#django.test.runner.DiscoverRunner) and a release note in `releases/4.0.txt` would be appropriate.

Just made these changes. Not sure if the docs and the release notes are under the appropriate heading, let me know if I need to make amends.

**Review by cjerdonek — COMMENTED** [review]
Some comments.

File: tests/test_runner/test_discover_runner.py
Comment (abbasidaniyal) [reply tests/test_runner/test_discover_runner.py:339]: I created a variable because it had to be checked in the `self.assertIn(msg, stdout.getvalue())` and 
`self.assertNotIn(msg, stdout.getvalue())`. Should I replace those occurences with a hard-coded string?

File: tests/test_runner/test_discover_runner.py
Comment (abbasidaniyal) [reply tests/test_runner/test_discover_runner.py:356]: Ohh I got it now. Got confused about the logging levels. My bad.

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:600]: > Some comments.

Thank you for taking time out the doing such an extensive review. I've incorporated the chanegs you suggested and also fixed up the logic.

File: tests/test_runner/test_discover_runner.py
Comment (cjerdonek) [reply tests/test_runner/test_discover_runner.py:339]: Yeah, in tests it's often okay to duplicate things. (By the way, can you just test equality of strings instead of contains? That would be more precise, and you would just have a single assert line with two different expected values.)

File: django/test/runner.py
Comment (cjerdonek) [reply django/test/runner.py:598]: No problem, happy to help. Thanks for keeping at it.

File: tests/test_runner/test_discover_runner.py
Comment (abbasidaniyal) [reply tests/test_runner/test_discover_runner.py:337]: Okay. Will make the changes

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:586]: No empty line at the beginning.

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:590]: This can be moved to the very beginning.

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:591]: You can simplify the rest of the method by doing:

```python
if self.verbosity == 1 and level is not None and level < logging.INFO:
    return
print(msg)
```

(This is what I meant when I said the `logging_level` variable could be eliminated.)

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:642]: It looks like you need to pass `level` `DEBUG` here and below.

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:735]: Again, it looks like you need to pass the level.

File: docs/topics/testing/advanced.txt
Comment (cjerdonek) [inline docs/topics/testing/advanced.txt:711]: Backticks are needed around `INFO`, `1`, `DEBUG`, and `2`.

File: tests/test_runner/test_discover_runner.py
Comment (cjerdonek) [inline tests/test_runner/test_discover_runner.py:355]: You can remove the empty line above and the two empty lines below.

File: tests/test_runner/test_discover_runner.py
Comment (cjerdonek) [inline tests/test_runner/test_discover_runner.py:358]: -> `with self.subTest(verbosity=verbosity, level=level):`

File: tests/test_runner/test_discover_runner.py
Comment (cjerdonek) [inline tests/test_runner/test_discover_runner.py:363]: You can just say `expected`. (It's clear from the next line.)

File: django/test/runner.py
Comment (cjerdonek) [inline django/test/runner.py:585]: Since this might not be obvious from the method implementation, it's probably worth spelling out with another sentence in the docstring: "A verbosity of 1 logs INFO (the default level) or above, and verbosity 2 logs all levels" (can go on a separate line).

**Review by cjerdonek — COMMENTED** [review]
Okay, last round of comments (I think?). They're more minor than before. Thanks again for keeping at it, @abbasidaniyal.

File: docs/topics/testing/advanced.txt
Comment (smithdc1) [reply docs/topics/testing/advanced.txt:711]: Please also add a `.. versionadded 4.0` directive. With a shirt note saying this method was added. 

https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features

File: django/test/runner.py
Comment (abbasidaniyal) [reply django/test/runner.py:591]: We can club the `self.verbosity == 0` with this as well. Something like this
```
def log(self, msg, level=None):
    """Log the given message at the given logging level."""
    if self.verbosity == 0 or (
        self.verbosity == 1 and level is not None and level < logging.INFO
    ):
        return
    print(msg)
```

Comment (cjerdonek) [reply django/test/runner.py:591]: Yeah, that works, too.

**Review by cjerdonek — APPROVED** [review]

File: django/test/runner.py
Comment (smithdc1) [inline django/test/runner.py:584]: Do we need any validation for level? What happens if I pass in say -1 or "1" into level?

File: tests/test_runner/test_discover_runner.py
Comment (smithdc1) [inline tests/test_runner/test_discover_runner.py:353]: Should there be a test for custom logging level?

File: docs/topics/testing/advanced.txt
Comment (smithdc1) [inline docs/topics/testing/advanced.txt:714]: I keep coming back to the docs here. I'm not sure if it is clear enough about what `level` should be (it was obvious once I looked at the tests). The link to the Python docs is great, but it's a massive page to expect folk to understand. I guess in part this ties in with one of the [comments on the ticket](https://code.djangoproject.com/ticket/32552#comment:6) about folk struggling with logging.

**Review by smithdc1 — COMMENTED** [review]
Your prompt for a review made me look at this again. :smile: 

A few small questions/comments, I suspect most can be ignored.

File: docs/topics/testing/advanced.txt
Comment (cjerdonek) [reply docs/topics/testing/advanced.txt:714]: The Django docs seem to prioritize brevity over completeness, but here is a possible longer version if more guidance is preferred. I would leave it up to the Django maintainers as to how much detail should be included:

> Prints to the console a message with the given integer [logging level](https://docs.python.org/3/library/logging.html#levels) (e.g. `logging.DEBUG`, `logging.INFO`, or `logging.WARNING`), respecting the current `verbosity`. For example, an ``INFO`` message will be logged if the verbosity is at least ``1``, and ``DEBUG`` will be logged if it is at least ``2``.

As an aside, this will make a lot more sense as a documented method once [ticket #32552](https://code.djangoproject.com/ticket/32552) is fully implemented. The reason is that the method is primarily useful for when a `logger` can be configured on the `DiscoverRunner` instance (in which case the user would presumably already be familiar with logging). The behavior of printing to the console will be the fallback behavior in that case, rather than the primary purpose for the method.

File: django/test/runner.py
Comment (felixxm) [reply django/test/runner.py:584]: Verbosity is already validated `django.core.management.base.BaseCommand`, I don't think we need more than that.

File: docs/topics/testing/advanced.txt
Comment (felixxm) [reply docs/topics/testing/advanced.txt:714]: I like Chris' proposition. I think it makes sense even without a logger in use.

File: tests/test_runner/test_discover_runner.py
Comment (felixxm) [reply tests/test_runner/test_discover_runner.py:353]: I added tests for custom levels.

**Comment by felixxm** [comment]
Thanks y'all :rocket: 

I rebased and pushed small edits.

**Pull request merged**
